core/src/avm2/verify.rs

```diff
@@ -561,279 +670,607 @@ fn optimize<'gc>(
+            }
+            Op::PushByte { value } => {
+                if *value >= 0 {
+                    stack.push_uint();
+                } else {
+                    stack.push_int();
+                }
+            }
+            Op::PushShort { value } => {
+                if *value >= 0 {
+                    stack.push_uint();
+                } else {
+                    stack.push_int();
+                }
+            }
+            Op::PushInt { value } => {
+                if *value < -(1 << 28) || *value >= (1 << 28) {
+                    stack.push_number();
+                } else if *value >= 0 {
+                    stack.push_uint();
+                } else {
+                    stack.push_int();
+                }
+            }
+            Op::DecrementI => {
+                // This doesn't give any Number-int guarantees
+                stack.pop();
+                stack.push_any();
+            }
+            Op::IncrementI => {
+                // This doesn't give any Number-int guarantees
+                stack.pop();
+                stack.push_any();
+            }
+            Op::DecLocalI { index } => {
+                if (*index as usize) < local_types.len() {
+                    // This doesn't give any Number-int guarantees
+                    local_types.set_any(*index as usize);
+                }
+            }
+            Op::IncLocalI { index } => {
+                if (*index as usize) < local_types.len() {
+                    // This doesn't give any Number-int guarantees
+                    local_types.set_any(*index as usize);
+                }
+            }
+            Op::Increment => {
+                stack.pop();
+                stack.push_number();
+            }
+            Op::Decrement => {
+                stack.pop();
+                stack.push_number();
+            }
+            Op::Negate => {
+                stack.pop();
+                stack.push_number();
+            }
+            Op::AddI => {
+                stack.pop();
+                stack.pop();
+                stack.push_any();
+            }
+            Op::SubtractI => {
+                stack.pop();
+                stack.pop();
+                stack.push_any();
+            }
+            Op::MultiplyI => {
+                stack.pop();
+                stack.pop();
+                stack.push_any();
+            }
+            Op::Add => {
+                stack.pop();
+                stack.pop();
+                stack.push_any();
+            }
+            Op::Subtract => {
+                stack.pop();
+                stack.pop();
+                stack.push_any();
+            }
+            Op::Multiply => {
+                stack.pop();
+                stack.pop();
+
+                // NOTE: In our current implementation, this is guaranteed,
+                // but it may not be after correctness fixes to match avmplus
+                stack.push_number();
```

Comment:
same as above: the fact that runtime value can be an int or float, doesn't mean that we can't still set it as being number-like.